### PR TITLE
[BugFix] Fix tool call finish reason in streaming case

### PR DESCRIPTION
**Pull request by maxdebayser**
When a named function is passed to tool_choice with streaming the finish_reason was "stop" instead of "tool_calls".

### CODE REVIEW

**Comment by maxdebayser** [comment]
cc: @njhill @tjohnson31415

File: vllm/entrypoints/openai/serving_chat.py [hunk 1]
```diff
@@ -539,11 +539,14 @@ async def chat_completion_stream_generator(
...
                         # only happens if we are NOT using guided decoding
                         if tool_parser:
-                            index = len(
-                                tool_parser.prev_tool_call_arr) - 1 if len(
-                                    tool_parser.prev_tool_call_arr) > 0 else 0
+                            tools_called = len(
+                                tool_parser.prev_tool_call_arr) > 0
+                            index = len(tool_parser.prev_tool_call_arr
+                                        ) - 1 if tools_called else 0
+                            tools_called = index > 0
```
Comment (tjohnson31415) [inline vllm/entrypoints/openai/serving_chat.py:546]: IMO, it is a bit confusing to set `tools_called` twice to different values.
```suggestion
                            index = len(
                                tool_parser.prev_tool_call_arr) - 1 if len(
                                    tool_parser.prev_tool_call_arr) > 0 else 0
                            tools_called = index > 0
```

**Comment by tjohnson31415** [comment]
Actually, this behavior may be expected in the OpenAI spec 🤔

> # Check if the model has made a tool_call. This is the case either if the "finish_reason" is "tool_calls" or if the "finish_reason" is "stop" and our API request had forced a function call
> ...
>     # This handles the edge case where if we forced the model to call one of our functions, the finish_reason will actually be "stop" instead of "tool_calls"

REF: https://platform.openai.com/docs/guides/function-calling/edge-cases

File: vllm/entrypoints/openai/serving_chat.py (hunk 1 above)
Comment (maxdebayser) [reply vllm/entrypoints/openai/serving_chat.py]: Oh thanks, that's actually a editing mistake resulting from working in a dev pod.

**Comment by maxdebayser** [comment]
> Actually, this behavior may be expected in the OpenAI spec 

Yeah, I'll test to see what OpenAI returns. If that is the case, then it's the non-streaming case that is inconsistent.

**Comment by maxdebayser** [comment]
@tjohnson31415 , you're right. I've tested with gpt4 on the OpenAI platform and the behavior is as follows:


OpenAI |   |  
-- | -- | --
  | named function | auto
unary | stop | tool_calls
stream | stop | tool_calls

So actually the non-streaming case for named function needs to be changed to return stop instead of tool_calls

File: vllm/entrypoints/openai/serving_chat.py [hunk 2]
```diff
@@ -703,7 +703,6 @@ async def chat_completion_full_generator(
                             name=request.tool_choice.function.name,
                             arguments=output.text))
                     ])
-                tools_called = True
```
Comment (tjohnson31415) [inline vllm/entrypoints/openai/serving_chat.py:706]: Can you add some comments about the behavior we are going for here to match the OpenAI spec?
The whole name and use of this `tools_called` variable may need some refactoring.

Comment (maxdebayser) [reply vllm/entrypoints/openai/serving_chat.py:706]: Yes, that makes sense

**Review by njhill — APPROVED** [review]
Thanks @maxdebayser!

**Pull request merged**